lib/LaTeXML/Core/Document.pm

```diff
@@ -1439,7 +1439,63 @@ sub pruneXMDuals {
       $self->collapseXMDual($dual, $presentation); }
     elsif (!$self->findnode('descendant-or-self::*[@_pvis or @_cvis]', $presentation)) {    # pres.
       $self->collapseXMDual($dual, $content); }
-  }
+    else {    # compact aligned structures, where possible
+      $self->compactXMDual($dual, $content, $presentation); } }
+  return; }
+
+sub compactXMDual {
+  my ($self, $dual, $content, $presentation) = @_;
+  # For now only handle compacting mirror applications
+  return if ($self->getNodeQName($content) ne 'ltx:XMApp') ||
+    ($self->getNodeQName($presentation) ne 'ltx:XMApp');
+  my @content_args = element_nodes($content);
+  my @pres_args    = element_nodes($presentation);
+  return if scalar(@content_args) != scalar(@pres_args);
+
+  my @new_args = ();
+  my $single_duality;
+  # walk the corresponding children, and double-check they are referenced in the same order
+  while ((my $c_arg = shift(@content_args)) and (my $p_arg = shift(@pres_args))) {
+    my $c_idref = $c_arg->getAttribute('idref');
+    if ($c_idref && ($c_idref eq ($p_arg->getAttribute('xml:id') || ''))) {
+      push @new_args, $p_arg;
+      next; }    # content-refs-pres, OK
+    my $p_idref = $p_arg->getAttribute('idref');
+    if ($p_idref && ($p_idref eq ($c_arg->getAttribute('xml:id') || ''))) {
+      push @new_args, $c_arg;
+      next; }    # pres-refs-content, OK
+
+    # Next up: differences. If:
+    # 1) we saw such a difference beforehand, or
+    # 2) the tree is too complex - give up on compacting and return.
+    # we only handle two XMToks differing for now.
+    if ($single_duality || ($self->getNodeQName($c_arg) ne 'ltx:XMTok') || $self->getNodeQName($p_arg) ne 'ltx:XMTok') {
+      return; }
+    else { # otherwise we can compact this case. but delay actual libxml changes until we are *sure* the entire tree is compactable
+      $single_duality = [$c_arg, $p_arg];
+      push(@new_args, $single_duality); } }
+  return unless $single_duality;
+
+# If we made it here, this is a dual with two mirrored applications and a single XMTok difference, compact it.
+  my $compact_apply = $self->openElementAt($dual->parentNode, 'ltx:XMApp');
+  for my $n_arg (@new_args) {
+    # one of the args has our dual node that needs compacting
+    if (ref $n_arg eq 'ARRAY') {
+      my ($c_arg, $p_arg) = @$n_arg;
+      # Transfer all c_arg attributes over, it should be primary?
+      for my $attr_key (qw(decl_id meaning name)) {
+        if (my $attr_val = $c_arg->getAttribute($attr_key)) {
+          $c_arg->removeAttribute($attr_key);
+          $p_arg->setAttribute($attr_key, $attr_val); } }
+      $n_arg = $p_arg; }
+    $n_arg->unbindNode;
+    $compact_apply->appendChild($n_arg); }
+  # if the dual has a role/id migrate them to the XMApp
+  for my $attr_key (qw(role xml:id)) {
```

Comment:
This lost an `lpadding` value in my case.  You need to copy more attributes from the XMDual to the new XMApp, not to mention from the old XMApp; probably all attributes (unless already set on the old XMApp).  And you'll want to use Document's setAttribute, since it manages ids.

It smells more&more like we need some sort of `$doc->mergeAttributes($oldnode,$newnode)` that knows which attributes to ignore, which for bookkeeping (id), which to append (class), and who knows else what crops up.

Reply:
Hm, alright. I could also rename the XMDual to an XMApp, that way no attribute moving needs to happen, I'll just replace its children.

Reply:
No, you'd still want to copy attributes from the old (presentation) XMApp.  (besides renaming nodes w/libxml2 always worries me; there've been problems in the past (and yeah, I know it's used rarely))

With the current approach all the XMRef's are in the content, so you really could just replace the XMDual with the presentation's XMApp (ie. an unwrap), still copying attributes from the XMDual down,  but I can see why you might not want to assume the current arrangement is permanent.

Reply:
Indeed. Since the schema is designed to allow XMRefs in either branch, might as well implement with the allowance in mind... I've added a commit that copies all XMDual attributes onto the new "compact" XMApp, unconditionally.

Reply:
Btw, I am not sure I can use the Document's `setAttribute`, as it reports an id conflict, since the id was already recorded on the now compacted XMDual. And we don't seem to have a Document `removeAttribute`, which removes the previous bookkeeping. Would this cause problems?

Reply:
Nevermind, figured it out, I first had to replace the node to un-document the xml:id, and then I could safely re-record it.

Reply:
No, but there's a `unRecordID`; it's a bit of dancing around, though.

Hmm, `setAttribute` doesn't use that when it *replaces* the id, though.